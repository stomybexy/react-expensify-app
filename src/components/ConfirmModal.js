import React from 'react';
import Modal from 'react-modal';

const ConfirmModal = (props) => (
    <Modal
        isOpen={props.isOpen}
        onRequestClose={props.onCancel}
        contentLabel={props.title}
        closeTimeoutMS={200}
        className="modal"
        ariaHideApp={false}
    >
        <p className="modal__body">{props.message || 'Do you confirm ?'}</p>

        <div className="modal__actions">
            <button id="no" className="button" onClick={props.onCancel}>
                No
            </button>
            <button id="yes" className="button  button--danger" onClick={props.onConfirm}>
                Yes
            </button>
        </div>
    </Modal>
);

export default ConfirmModal;