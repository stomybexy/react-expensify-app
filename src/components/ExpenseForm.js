import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';

export default class ExpenseForm extends React.Component {

    constructor(props) {
        super(props);
        let state = {
            description: '',
            note: '',
            amount: '',
            createdAt: moment(),
            calendarFocused: false,
            error: ''
        };
        if (props.expense) {
            state = {
                ...state,
                description: props.expense.description,
                note: props.expense.note,
                amount: (props.expense.amount / 100).toString(),
                createdAt: moment(props.expense.createdAt),
            }
        }
        this.state = state;
    }

    onDescriptionChange = ({target}) => this.setState(() => ({description: target.value}));

    onNoteChange = ({target}) => this.setState(() => ({note: target.value}));

    onAmountChange = ({target}) => {
        const amount = target.value;

        if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
            this.setState(() => ({amount}))
        }
    };

    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({createdAt}));
        }
    };

    onCalendarFocusChange = ({focused}) => this.setState({calendarFocused: focused});

    onSubmit = (e) => {
        e.preventDefault();

        const {description, amount, createdAt, note} = this.state;

        if (!description || !amount) {
            const error = 'Please provide description and amount.';
            this.setState(() => ({error}))
        } else {
            this.setState(() => ({error: ''}));
            this.props.onSubmit({
                description,
                amount: parseFloat(amount) * 100,
                createdAt: createdAt.valueOf(),
                note
            })
        }
    };

    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onCalendarFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    >
                    </textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        );
    }
}