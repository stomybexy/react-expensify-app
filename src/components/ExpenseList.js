import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getVisibleExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        <h1>Expense List</h1>
        {
            props.expenses.length === 0 ? (
                <p>No expenses</p>
            ) : (
                props.expenses.map(expense => (
                    <ExpenseListItem key={expense.id} {...expense}/>
                ))
            )
        }

    </div>
);

const mapStateToProps = ({expenses, filters}) => ({
    expenses: getVisibleExpenses(expenses, filters)
});

export default connect(mapStateToProps)(ExpenseList);
