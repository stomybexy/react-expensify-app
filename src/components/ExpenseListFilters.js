import React from 'react';
import {connect} from 'react-redux';
import {DateRangePicker} from 'react-dates';
import {setEndDate, setStartDate, setTextFilter, sortByAmount, sortByDate} from '../actions/filters';


export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    };

    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = calendarFocused => this.setState({calendarFocused});

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };

    onSortChange = ({target}) => {
        const sortBy = target.value;
        if (sortBy === 'date') {
            this.props.sortByDate();
        } else if (sortBy === 'amount') {
            this.props.sortByAmount();
        }
    };

    render() {
        return (
            <div>
                <input type="text" value={this.props.filters.text} onChange={this.onTextChange}/>
                <select
                    value={this.props.filters.sortBy}
                    onChange={this.onSortChange}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>

                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
                    showClearDates={true}
                    onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
                    focusedInput={this.state.calendarFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                    onFocusChange={this.onFocusChange} // PropTypes.func.isRequired,
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />

            </div>
        );
    }
}


const mapStateToProps = ({filters}) => ({filters});

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);