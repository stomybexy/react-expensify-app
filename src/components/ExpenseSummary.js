import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import getExpensesTotal from '../selectors/expenses-total';
import getVisibleExpenses from '../selectors/expenses';

export const ExpenseSummary = ({ expensesCount, expensesTotal, hiddenExpensesCount }) => {
    const expenseWord = expensesCount === 1 ? 'expense' : 'expenses';
    const hiddenExpenseWord = hiddenExpensesCount === 1 ? 'expense' : 'expenses';
    const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00');
    return (
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title">
                    Viewing <span>{expensesCount}</span> {expenseWord} totaling <span>{formattedExpensesTotal}</span>
                </h1>
                {!!hiddenExpensesCount && (
                    <h3 className="page-header__subtitle">{hiddenExpensesCount} hidden {hiddenExpenseWord} because of filters</h3>
                )}
                <div className="page-header__actions">
                    <Link className="button" to="/create">Add Expense</Link>
                </div>
            </div>
        </div>
    );
};
const mapStateToProps = ({ expenses, filters }) => {

    const visibleExpenses = getVisibleExpenses(expenses, filters);

    return {
        expensesCount: visibleExpenses.length,
        expensesTotal: getExpensesTotal(visibleExpenses),
        hiddenExpensesCount: expenses.length - visibleExpenses.length
    }
};
export default connect(mapStateToProps)(ExpenseSummary);