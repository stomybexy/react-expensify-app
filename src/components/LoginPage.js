import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ({ startLogin }) => (
    <div className="box-layout">
        <div className="box-layout__box">
            <h1 className="box-layout__title">Expensify</h1>
            <p>It's time to get your expenses under control</p>
            <div className="box-layout__actions">
                <button className="button" onClick={() => startLogin('GOOGLE')}>
                    Login with Google
                </button>
                <button className="button" onClick={() => startLogin('GITHUB')}>
                    Login with Github
                </button>
            </div>
        </div>
    </div>
);

const mapDispatchToprops = (dispatch) => ( {
    startLogin: (providerName) => dispatch(startLogin(providerName))
} );

export default connect(undefined, mapDispatchToprops)(LoginPage);