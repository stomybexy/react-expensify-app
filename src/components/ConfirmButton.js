import React, { Component } from 'react';
import ConfirmModal from './ConfirmModal';

export default class ConfirmButton extends Component {

    state = {
        modalOpen: false
    };

    requestConfirm = () => this.setState({ modalOpen: true });

    onCancel = () => this.setState({ modalOpen: false });

    render() {
        return (
            <div>
                <button
                    className={this.props.className}
                    onClick={this.requestConfirm}>
                    {this.props.children}
                </button>

                <ConfirmModal
                    title={this.props.confirmTitle}
                    message={this.props.confirmMessage}
                    isOpen={this.state.modalOpen}
                    onConfirm={this.props.onConfirm}
                    onCancel={this.onCancel}
                />
            </div>
        );

    }
}
