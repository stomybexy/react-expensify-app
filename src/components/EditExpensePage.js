import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import ConfirmButton from './ConfirmButton';

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    };
    onRemoveButtonClick = () => {
        this.props.startRemoveExpense({ id: this.props.expense.id });
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm
                        expense={this.props.expense}
                        onSubmit={this.onSubmit}
                    />
                    <ConfirmButton
                        className="button button--secondary"
                        onConfirm={this.onRemoveButtonClick}
                        confirmTitle="Confirm removal"
                        confirmMessage="Remove this expense ?"
                    >
                        Remove Expense
                    </ConfirmButton>
                </div>
            </div>
        );
    }
}

const mapStateToProps = ({ expenses }, props) => ( {
    expense: expenses.find(({ id }) => id === props.match.params.id)
} );

const mapDispatchToProps = (dispatch) => ( {
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
} );

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
