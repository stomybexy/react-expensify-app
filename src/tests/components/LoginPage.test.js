import React from 'react';
import { shallow } from 'enzyme';
import { LoginPage } from '../../components/LoginPage';

let wrapper, startLogin;

beforeEach(() => {
    startLogin = jest.fn();
    wrapper = shallow(<LoginPage startLogin={startLogin}/>);
});

test('should render correctly LoginPage', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should call startLogin with google on google button click', () => {
    wrapper.find('#google').simulate('click');
    expect(startLogin).toHaveBeenLastCalledWith('GOOGLE');
});
test('should call startLogin with github on github button click', () => {
    wrapper.find('#github').simulate('click');
    expect(startLogin).toHaveBeenLastCalledWith('GITHUB');
});