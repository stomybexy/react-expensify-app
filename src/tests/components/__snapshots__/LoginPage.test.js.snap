// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly LoginPage 1`] = `
<div
  className="box-layout"
>
  <div
    className="box-layout__box"
  >
    <h1
      className="box-layout__title"
    >
      Expensify
    </h1>
    <p>
      It's time to get your expenses under control
    </p>
    <div
      className="box-layout__actions"
    >
      <button
        className="button"
        id="google"
        onClick={[Function]}
      >
        Login with Google
      </button>
      <button
        className="button"
        id="github"
        onClick={[Function]}
      >
        Login with Github
      </button>
    </div>
  </div>
</div>
`;
